pcfuncs/ipban/models.py

```diff
@@ -0,0 +1,71 @@
+import logging
+from typing import Any, List, Set
+
+from azure.data.tables import TableClient, UpdateMode
+from azure.monitor.query import LogsQueryClient
+from azure.monitor.query._models import LogsTableRow
+
+from .constants import (
+    LOG_ANALYTICS_WORKSPACE_ID,
+    THRESHOLD_READ_COUNT_IN_GB,
+    TIME_WINDOW_IN_HOURS,
+)
+
+
+class UpdateBannedIPTask:
+    def __init__(
+        self,
+        logs_query_client: LogsQueryClient,
+        table_client: TableClient,
+    ) -> None:
+        self.log_query_client: LogsQueryClient = logs_query_client
+        self.table_client: TableClient = table_client
+
+    def run(self) -> List[LogsTableRow]:
+        query_result: List[LogsTableRow] = self.get_blob_logs_query_result()
+        self.update_banned_ips(query_result)
+        return query_result
+
+    def get_blob_logs_query_result(self) -> List[LogsTableRow]:
+        query: str = f"""
+        StorageBlobLogs
+        | where TimeGenerated > ago({TIME_WINDOW_IN_HOURS}h)
+        | extend IpAddress = tostring(split(CallerIpAddress, ":")[0])
+        | summarize readcount = sum(ResponseBodySize) / (1024 * 1024 * 1024)
+        by IpAddress
+        | where readcount > {THRESHOLD_READ_COUNT_IN_GB}
+        """
+        response: Any = self.log_query_client.query_workspace(
+            LOG_ANALYTICS_WORKSPACE_ID, query, timespan=None
+        )
+        return response.tables[0].rows
+
+    def update_banned_ips(self, query_result: List[LogsTableRow]) -> None:
```

Comment:
Curious about this implmentation a little. If all of these lists are small, doing 3 loops sequentially won't be an issue. However, is there a benefit to just truncating the table and inserting? I'm guessing this approach keeps banned IPs that will continue to be banned in the table, such that they are not momentarily "un-banned". That works nicely; I guess I'd mostly like some indication that the list sizes here won't make this an issue.

Reply:
This list will be short, so I prefer this approach as it ensures higher availability. It is crucial to always have a list of banned IPs for the application to reference, preventing egress abuse.